### PR TITLE
Fix Sight Line for players/monsters

### PR DESCRIPTION
**Pull request by gigastar**
Fixed issues with #3223 by zbizu.

### CODE REVIEW

**Comment by MillhioreBT** [comment]
What is this about? I was reading the references and such an error cannot be reproduced in my sources? Is it an improvement of the code or are you really fixing something essential?

**Comment by gigastar** [comment]
It's possible the problems I found were linked to my set-up.

Pretty much the sight line worked correctly (if you look at the picture where gold coins were being placed) #3223.

It was working correctly like that, however you couldn't do the same thing to the north, and it was completely breaking if you tried it to the left rather than the right.

With the changes I made, in my set up, made it so the sight line is the same in all directions. Which it should be. There may be something I have overseen so it does need to be tested more.

I would suggest you do some testing to figure out if this has changed anything.

**Comment by MillhioreBT** [comment]
> It's possible the problems I found were linked to my set-up.
> 
> Pretty much the sight line worked correctly (if you look at the picture where gold coins were being placed) #3223.
> 
> It was working correctly like that, however you couldn't do the same thing to the north, and it was completely breaking if you tried it to the left rather than the right.
> 
> With the changes I made, in my set up, made it so the sight line is the same in all directions. Which it should be. There may be something I have overseen so it does need to be tested more.
> 
> I would suggest you do some testing to figure out if this has changed anything.

Horizontally it works fine, however vertically it cannot be verified with the official TibiaClient, due to limited vision, however I verified it using the internalMoveItem function manually and it also works correctly.

I apologize if I still do not understand your point, am I doing something wrong?

image
image
image
image

**Comment by gigastar** [comment]
Unless i'm mistaken you should be able to throw in the first 2 spots after the walls corner (if you are 3sqm away from the corner)
Which is represented in your first picture.

**Comment by gigastar** [comment]
So this is the problem. There is also a problem in this code I posted now that I have run some more tests..

The gold coins represent here you can throw
The platinum represent where you cannot (but should be able to)

This is the outcome of Zbizu's code
show2


This is the outcome of the code I have modified. (I am working on fixing it now)
show1

**Comment by gigastar** [comment]
This is the results of changes I just made. The only issue right now is that the top left isn't working 100%.
show1

**Comment by MillhioreBT** [comment]
The question is ... this is so in TibiaRL?

because in this image you can see how the red lines collide with the wall
<details>
  <summary>Images:</summary>

image
image

</details>

**Comment by gigastar** [comment]
True and I was assuming the information from zbizu and whoever else talking about that issue was correct about how it should be to mirror rl tibia. I am pretty sure that was the correct way. Are you using stock TFS sources?

**Comment by MillhioreBT** [comment]
> True and I was assuming the information from zbizu and whoever else talking about that issue was correct about how it should be to mirror rl tibia. I am pretty sure that was the correct way. Are you using stock TFS sources?

I have checked TFS is wrong, but its configuration is also wrong, in TibiaRL you can only do the following:

<details>
<summary>Images:</summary>

image
image

</details>

**Comment by gigastar** [comment]
Are you saying you shouldn't be able to throw the same to the north and to the west?

**Comment by MillhioreBT** [comment]
> Are you saying you shouldn't be able to throw the same to the north and to the west?

Correct, I confirmed it in TibiaRL
You can also try to enter TibiaRL and check it

**Comment by gigastar** [comment]
Well that doesn't make sense but okay.

**Comment by MillhioreBT** [comment]
> Well that doesn't make sense but okay.

In an isometric world it makes a lot of sense, although it does not seem so for a flat or three-dimensional world, I also think the same as you, but this repository is governed by the mechanics of TibiaRL

**Comment by gigastar** [comment]
This is the end result. I was able to reproduce the bug with the extra spot being thrown on the right side. An offset was added to the grade of a line which was needed to reproduce the isometric way the line should be drawn. I did test this in a vertical layout as-well.
show

**Comment by MillhioreBT** [comment]
> This is the end result. I was able to reproduce the bug with the extra spot being thrown on the right side. An offset was added to the grade of a line which was needed to reproduce the isometric way the line should be drawn. I did test this in a vertical layout as-well.
> show

changes are 99% similar to TibiaRL
only this tile that I'm pointing with the green arrow must be unreachable
image

I'm sorry if I confused you with my images
great job fix what i told you and hopefully someone else approves it really works just like TibiaRL

**Comment by gigastar** [comment]
That tile is unreachable on my side, is it reachable on your set up?

**Comment by MillhioreBT** [comment]
> That tile is unreachable on my side, is it reachable on your set up?

yes :/ 
image

**Comment by MillhioreBT** [comment]
> That tile is unreachable on my side, is it reachable on your set up?

yes :/

**Comment by gigastar** [comment]
That wall isn't completed. You need the last end part (the pillar)

**Comment by nekiro** [comment]
What was the tested scenarious?
I believe this also affects monsters targetung right? Was it thoroughly tested?

**Comment by MillhioreBT** [comment]
> What was the tested scenarios?
> I believe this also affects monsters targeting and rune throwing (also other spells possibly?) right? Was it thoroughly tested?

I have already tried all the scenarios, and the monsters comply with the proposed rules, here I will leave the image provided by @EPuncker from Tibia RL

As you can see, the orc that is to the north is unreachable and therefore his strange walk to get closer


What still seems strange to me is that in my sources, these changes seem to err a bit, because I can move objects to the following place:


If anyone else can check that that would be great

**Comment by gigastar** [comment]
Okay, now I am seeing the problem there. Thank you for slapping me a couple times. Ill get that fixed.

**Comment by gigastar** [comment]
Try now please

**Comment by MillhioreBT** [comment]
Meets expectations 😍
GIF 08-01-2021 11-09-46 p  m

Now if it perfectly emulates the behavior of Tibia RL

**Review by MillhioreBT — APPROVED** [review]

**Comment by gigastar** [comment]
To be clear all credits to Zbizu he did great implementing the Xiaolin Wu algorithm to it. I just added an offset that was needed. my dumb a** almost made it so it works in all directions but we can forget about that.

File: src/map.cpp
Comment (ranisalt) [inline src/map.cpp:536]: What algorithm is this? Please link to it in the code itself.

File: src/map.cpp
Comment (ranisalt) [inline src/map.cpp:539]: Remove the spaces after < and before > on both lines

File: src/map.cpp
Comment (ranisalt) [inline src/map.cpp:554]: Tip: you can use [std::minmax](https://en.cppreference.com/w/cpp/algorithm/minmax) to achieve the same result without the explicit branch:

```cpp
std::tie(x0, x1) = std::minmax(x0, x1);
std::tie(y0, y1) = std::minmax(y0, y1);
```
And then `x0` will always be less or equal than `x1`, ditto for `y0` and `y1`.

File: src/map.cpp
Comment (ranisalt) [inline src/map.cpp:563]: Where does `0.03` come from?

File: src/map.cpp
Comment (ranisalt) [inline src/map.cpp:580]: There will be speed benefits if you `reserve` space in `line` before `push_back`s. It will always be `x1 - x0` long. Also, use `emplace_back` instead.

By the way, `++x` is preferred over `x++`.

File: src/map.cpp
Comment (ranisalt) [inline src/map.cpp:590]: Instead of this `static_cast` cluster, make your `get2DLine` function take the correct types.

File: src/map.cpp
Comment (ranisalt) [inline src/map.cpp:629]: Spacing

File: src/map.cpp
Comment (ranisalt) [inline src/map.cpp:676]: This is a `for` loop, don't use `while`.

**Review by ranisalt — CHANGES_REQUESTED** [review]
Most of your mistakes are related to not following the code convention. Please use as much as possible of the current code style.

File: src/map.cpp
Comment (Zbizu) [reply src/map.cpp:536]: I was basing that code on this:
https://rosettacode.org/wiki/Xiaolin_Wu%27s_line_algorithm#C.2B.2B

File: src/map.cpp
Comment (Zbizu) [reply src/map.cpp:590]: agreed. I used this to avoid issues with converting data types (like #3217), but it should definitely be rewritten.

File: src/map.cpp
Comment (nekiro) [reply src/map.cpp:571]: Probably from testing the code and finding the right value, but it's just my wild guess.

**Review by gigastar — COMMENTED** [review]
Yes, it was after a lot of researching the different grad values that were being used in each spot. What I gathered was the issue was taking place if the grad was: 0.3333 repeating. If the grad was anything else it was fine. That value is just enough that it won't interfere with the way the line is reading in any other scenario. It is hard to explain without you seeing all grad values I printed out after trying to throw a coin in every spot open.

It's definitely not optimal. A better way would be to add the offset only in the specific instance its needed. This is what I was able to come up with as an alternative. Not that its detrimental to anything.

File: src/map.cpp
Comment (gigastar) [reply src/map.cpp:571]: > Probably from testing the code and finding the right value, but it's just my wild guess.

Yes, it was after a lot of researching the different grad values that were being used in each spot. What I gathered was the issue was taking place if the grad was: 0.3333 repeating. If the grad was anything else it was fine. That value is just enough that it won't interfere with the way the line is reading in any other scenario. It is hard to explain without you seeing all grad values I printed out after trying to throw a coin in every spot open.

It's definitely not optimal. A better way would be to add the offset only in the specific instance its needed. This is what I was able to come up with as an alternative. Not that its detrimental to anything.

Comment (gigastar) [reply src/map.cpp:571]: It would be good for someone to check it with otclient and see if the line remains the same past the limits of the tibia client but that's not what TFS is for.

File: src/map.cpp
Comment (gigastar) [reply src/map.cpp:560]: I tried to implement this, it could be because i'm in over my head with it, but it wouldn't produce the same outcome.

File: src/map.cpp
Comment (ranisalt) [reply src/map.cpp:571]: I think the ideal solution would be to mirror the vector to always be in the same direction, then calculate only one variation i.e. normalize every distance to the positive vector (top-left to bortom-right) and check that without magic numbers.

Comment (Zbizu) [reply src/map.cpp]: I recommend testing this on both Linux and Windows anyway because they handle calculations a bit differently (for example math.random gets different results on same seed).

File: src/map.cpp
Comment (Zbizu) [inline src/map.cpp:588]: outdated comment referencing removed lines of code

File: src/map.cpp
Comment (Zbizu) [inline src/map.cpp:618]: Another outdated comment. New algorithm doesn't cast two rays.

**Review by Zbizu — CHANGES_REQUESTED** [review]

**Comment by Zbizu** [comment]
> 
> 
> To be clear all credits to Zbizu he did great implementing the Xiaolin Wu algorithm to it. I just added an offset that was needed. my dumb a** almost made it so it works in all directions but we can forget about that.

Both you and ranisalt did a great job getting the code to tfs standards. I can have great algorhytmical thinking but I'm still a noob at c++.

**Comment by gigastar** [comment]
@ranisalt do I HAVE to change anything else?

**Review by ranisalt — DISMISSED** [review]
The code is now conforming to the standard, though I can't test the behavior. I'll trust @MillhioreBT on that.

**Comment by nekiro** [comment]
I'll give this a test soon, need to find some free time.

**Comment by nekiro** [comment]
Did anyone check performance of this code comparing to the old code?

**Comment by EPuncker** [comment]
ok, so I tested these changes on my testing tfs (not really clean code, but close to master) and the monsters started behaving weirdly, so I'd like to ask you guys to test how are monsters behaving (with normal characters)

**Comment by gigastar** [comment]
Tested on base files and everything seems to be normal. I tried all 3 scenarios that are in the original issue. Also some other ones just to see if anything weird was going on. I can't see any problem other than the slight differences in monster's AI behavior compared to Tibia.

**Comment by gigastar** [comment]
After doing some more testing there seems to be a problem with the code. If someone can test it to make sure. For some reason items are not able to be thrown correctly. Which is weird because it was working fine before.

**Comment by gigastar** [comment]
After doing more testing I found a problem with throwing items. It was due to trying to use const_cast and modifying the Position& created. I had to revert to x0,x1, ect. After those changes I tested all scenarios again and everything seems to be correct.

**Comment by emil92b** [comment]
try using spells nearby or trough magic wall

**Comment by gigastar** [comment]
Not sure how that got changed

**Comment by emil92b** [comment]
should only be able to throw items up 1 floor, and down 7 floors

see image: https://i.imgur.com/oyrjmls.png (no limit)

currently you can only throw down 2 floors

**Comment by gigastar** [comment]
The issue with how many floors down you can throw isn't related to this commit. I will look into being able to throw through magic walls.

**Comment by gigastar** [comment]
Anyone have any ideas on the magic wall issue? I tried using Field* field = tile->getFieldItem() and checking if its a blocksolid, ect, ect. Nothing I have tried seems to work. I did fix the item throw range down floors but that will be in a different commit.

**Comment by Zbizu** [comment]
Do magic walls have blockProjectile in otb?

**Comment by nekiro** [comment]
> Do temporary magic walls have blockProjectile in otb?

They are immovable blocksolids by default.

**Comment by EPuncker** [comment]
> > Do temporary magic walls have blockProjectile in otb?
> 
> They are immovable blocksolids by default.

I think @Zbizu means if magic wall have <attribute key="blocking" value="1" /> on otb

currently our items.xml is like that:

```
	<item id="1497" article="a" name="magic wall">
		<attribute key="type" value="magicfield" />
		<attribute key="decayTo" value="0" />
		<attribute key="duration" value="20" />
	</item>
	<item id="1499" article="a" name="rush wood">
		<attribute key="type" value="magicfield" />
		<attribute key="decayTo" value="0" />
		<attribute key="duration" value="45" />
		<attribute key="blocking" value="1" />
	</item>
```
But I don't know if that match 10.98 data as I don't have it to test

**Comment by Zbizu** [comment]
`blocksolid` and `blockprojectile` are different flags. First tells that the object is something like a wall, or some rock (they get replaced by default in map editor unless you stack them on top of each other through moving), second tells that you can't shoot through it. If you have a window it's an object you can't walk through, but you can shoot through it.

https://github.com/otland/forgottenserver/blob/master/src/items.cpp#L447

**Comment by gigastar** [comment]
I know I was just saying I tried looking for all flags and none of them changed the outcome.

**Comment by Zbizu** [comment]
have you tried obtaining the magic wall by id? tile->getFieldItem might be bugged

I remember that it was possible to get mw by id back in tfs 1.1:
https://otland.net/threads/10-76-tfs-1-1-old-mw-sprite-timer.230810/

**Comment by gigastar** [comment]
So you can throw through any walls with runes, ect. The problem is bigger than just mwalls.

**Comment by gigastar** [comment]
Last commit should fix the problem.

**Comment by emil92b** [comment]
> Last commit should fix the problem.

it breaks item throwing like this tho, maby something else too
https://i.imgur.com/k6ee1rt.png

**Comment by gigastar** [comment]
Couldn't reproduce that.

**Comment by nekiro** [comment]
> Couldn't reproduce that.

He means throwing over the roof.

**Comment by gigastar** [comment]
I don't understand.

**Comment by Zbizu** [comment]
I coded it to ignore borders:
`bool obstacleFound = tile && (tile->getGround() || tile->hasProperty(CONST_PROP_BLOCKPROJECTILE));`

Current tfs implementation for that is:
https://github.com/otland/forgottenserver/blob/master/src/map.cpp#L565

reason for my change: I tried to throw coins from tfs rook temple roof the the floor and I wasn't able to throw them next to temple wall tiles because the shaft was covered by roof border. If I remember correctly this doesn't happen in rl.

**Comment by gigastar** [comment]
So is there a problem or? Can someone please explain a bit better im a little slow ;)

**Comment by Zbizu** [comment]
You stand on a roof of a building. If the roof has a border, tfs prevents you to throw through it so ground floor tiles near building walls become unreachable.

**Comment by gigastar** [comment]
You can throw passed roof boarders on rl tibia, yes.

**Comment by gigastar** [comment]
It seems you can throw over walls/roofs if its 1 floor also which can't be done on TFS

**Comment by gigastar** [comment]
You can throw past borders on RL tibia.

I have made it so we can determine what is blocking the sightline in the path. If it is a magic wall then we do not allow throwing over it. This should fix the last couple issues brought up. Now we need to test more.

**Comment by Zbizu** [comment]
oops, linked wrong line earlier, here's the one responsible for checking if you can throw through floors in current tfs:
https://github.com/otland/forgottenserver/blob/master/src/map.cpp#L573

**Comment by gigastar** [comment]
TFS and RL tibia are both able to move items through borders. I fixed the problem mentioned by emil92b, I believe.

File: src/map.cpp
Comment (EPuncker) [inline src/map.cpp:652]: } else {

**Review by EPuncker — CHANGES_REQUESTED** [review]
this need a rebase tho

File: src/map.cpp
Comment (Zbizu) [inline src/map.cpp:603]: one tab too much in this line

**Comment by gigastar** [comment]
How does the code now deviate from that?

**Comment by nekiro** [comment]
@TinkzBot that's definitely great information, can you please post your research in #2891?

**Comment by Zbizu** [comment]
I see this was closed and the repository having these changes isn't reachable anymore. Are you going to continue working on that code @gigastar ?

What needs to be done to finish it? Only interaction with magic wall fix?

**Comment by gigastar** [comment]
I am not sure if anything needs to be added for it to be finished. I fixed all the problems that were mentioned. The code wasn't beautiful but it wasn't terrible.

1) You can throw over 1 floor
2) Problem throwing over magic walls was fixed (You can not now.)
3) Magic walls blocked throwing if found anywhere in the path
4) The sightline seemed to mirror tibias sightline by all testing I did
5) Just waiting for reviews on code changes (to mimic TFS style) and any testing anyone wants to do the make sure the sightline is correct and everything previously mentioned is correct.

**Comment by MillhioreBT** [comment]
@gigastar your commitment not only improves behavior, but emulates the real behavior of Tibia and fixes other existing problems, if your commitment was not accepted quickly it is due to the fact that this function is very delicate and is present in almost all the mechanics of the game and it should be reviewed very well before merging it, I hope you are not discouraged and at some point re-open this PR, you do not need to close it because we do not accept it immediately

**Pull request closed without merging**
